In [1]:
# part1

In [1]:
import pandas as pd 
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
pd.__version__ 

'2.3.3'

In [4]:
df = pd.read_csv("C:\\Users\\dell\\Desktop\\anaconda\\Pima Indians Diabetes.csv") 

In [5]:
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
...,...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1


In [6]:
df = df.sample(frac=1,random_state = 20) 
df = df.reset_index(drop=True) 

In [7]:
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
1,5,166,72,19,175,25.8,0.587,51,1
2,3,96,78,39,0,37.3,0.238,40,0
3,8,110,76,0,0,27.8,0.237,58,0
4,0,126,86,27,120,27.4,0.515,21,0
...,...,...,...,...,...,...,...,...,...
763,5,85,74,22,0,29.0,1.224,32,1
764,7,142,60,33,190,28.8,0.687,61,0
765,2,108,62,32,56,25.2,0.128,21,0
766,4,114,64,0,0,28.9,0.126,24,0


In [8]:
df_train = df.iloc[:614].copy(deep =True)
df_val = df.iloc[614:691].copy(deep =True).reset_index(drop=True) 
df_test = df.iloc[691:].copy(deep =True).reset_index(drop=True) 

In [9]:
print(df_train.shape , df_val.shape , df_test.shape)

(614, 9) (77, 9) (77, 9)


In [10]:
print(df.shape) 

(768, 9)


#8
Pregnancies: تعداد دفعات بارداری این ستون نشان می‌دهد که یک زن چند بار باردار شده است.
Glucose: سطح گلوکز خون (قند خون). این اندازه‌گیری معمولاً پس از ناشتا بودن انجام می‌شود و می‌تواند نشان‌دهنده مشکلاتی مانند دیابت باشد.
BloodPressure: فشار خون دیاستولیک (عدد پایینی فشار خون). این مقدار بر حسب میلی‌متر جیوه (mmHg) است.
SkinThickness: ضخامت چین پوستی سه سر بازو. این اندازه‌گیری می‌تواند به تخمین میزان چربی بدن کمک کند.
Insulin: سطح انسولین سرم. انسولین هورمونی است که به تنظیم قند خون کمک می‌کند.
BMI: شاخص توده بدنی. این شاخص با تقسیم وزن (به کیلوگرم) بر مجذور قد (به متر) محاسبه می‌شود و معیاری برای ارزیابی وزن نسبت به قد است.
DiabetesPedigreeFunction: تابع شجره‌نامه دیابت. این یک معیار ژنتیکی است که احتمال ابتلا به دیابت را بر اساس سابقه خانوادگی فرد ارزیابی می‌کند. مقدار بالاتر نشان‌دهنده ریسک بالاتر است.
Age: سن فرد (بر حسب سال).
Outcome: نتیجه تشخیص. این ستون نشان می‌دهد که آیا فرد به دیابت مبتلا است یا خیر. مقدار ‘0’ به معنی عدم ابتلا به دیابت و مقدار ‘1’ به معنی ابتلا به دیابت است.

In [11]:
df_train.head() 

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
1,5,166,72,19,175,25.8,0.587,51,1
2,3,96,78,39,0,37.3,0.238,40,0
3,8,110,76,0,0,27.8,0.237,58,0
4,0,126,86,27,120,27.4,0.515,21,0


In [12]:
df_train.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               614 non-null    int64  
 1   Glucose                   614 non-null    int64  
 2   BloodPressure             614 non-null    int64  
 3   SkinThickness             614 non-null    int64  
 4   Insulin                   614 non-null    int64  
 5   BMI                       614 non-null    float64
 6   DiabetesPedigreeFunction  614 non-null    float64
 7   Age                       614 non-null    int64  
 8   Outcome                   614 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 43.3 KB


اولین اندیس:1
آخرین اندیس: 768
تعداد کل ردیف‌ها: 768

In [13]:
df_train.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000
mean,3.724756,121.161238,69.087948,20.750814,81.701954,32.113192,0.474873,32.874593,0.366450
std,3.301708,32.848335,19.830326,16.007180,115.837291,7.863379,0.336785,11.437278,0.482227
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,64.000000,0.000000,0.000000,27.400000,0.245250,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,36.000000,32.350000,0.375500,29.000000,0.000000
75%,6.000000,141.000000,80.000000,32.000000,130.000000,36.600000,0.621500,40.000000,1.000000
max,15.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,72.000000,1.000000


In [14]:
df.head(10) 

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
1,5,166,72,19,175,25.8,0.587,51,1
2,3,96,78,39,0,37.3,0.238,40,0
3,8,110,76,0,0,27.8,0.237,58,0
4,0,126,86,27,120,27.4,0.515,21,0
5,5,109,62,41,129,35.8,0.514,25,1
6,5,97,76,27,0,35.6,0.378,52,1
7,1,140,74,26,180,24.1,0.828,23,0
8,2,128,64,42,0,40.0,1.101,24,0
9,7,81,78,40,48,46.7,0.261,42,0


In [15]:
df.tail(3)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
765,2,108,62,32,56,25.2,0.128,21,0
766,4,114,64,0,0,28.9,0.126,24,0
767,9,165,88,0,0,30.4,0.302,49,1


In [16]:
df.iloc[:, 4:7]

,Insulin,BMI,DiabetesPedigreeFunction
0,0,32.4,0.272
1,175,25.8,0.587
2,0,37.3,0.238
3,0,27.8,0.237
4,120,27.4,0.515
...,...,...,...
763,0,29.0,1.224
764,190,28.8,0.687
765,56,25.2,0.128
766,0,28.9,0.126


In [17]:
df_train.iloc[:20]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
1,5,166,72,19,175,25.8,0.587,51,1
2,3,96,78,39,0,37.3,0.238,40,0
3,8,110,76,0,0,27.8,0.237,58,0
4,0,126,86,27,120,27.4,0.515,21,0
5,5,109,62,41,129,35.8,0.514,25,1
6,5,97,76,27,0,35.6,0.378,52,1
7,1,140,74,26,180,24.1,0.828,23,0
8,2,128,64,42,0,40.0,1.101,24,0
9,7,81,78,40,48,46.7,0.261,42,0


In [18]:
df.iloc[14:35, 1:5]

,Glucose,BloodPressure,SkinThickness,Insulin
14,106,64,35,119
15,72,78,25,0
16,97,64,19,82
17,102,64,46,78
18,136,74,50,204
19,115,0,0,0
20,108,62,10,278
21,106,60,24,0
22,130,64,0,0
23,130,70,13,105


In [19]:
df_train["Pregnancies"] 

0      10
1       5
2       3
3       8
4       0
       ..
609     3
610     2
611     7
612    12
613     4
Name: Pregnancies, Length: 614, dtype: int64

In [20]:
df_train[["BMI","Insulin","Glucose"]] 

,BMI,Insulin,Glucose
0,32.4,0,108
1,25.8,175,166
2,37.3,0,96
3,27.8,0,110
4,27.4,120,126
...,...,...,...
609,24.7,115,96
610,21.1,49,96
611,37.2,105,102
612,30.0,105,100


In [21]:
df_train.loc[df_train["Age"] <= 50]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
2,3,96,78,39,0,37.3,0.238,40,0
4,0,126,86,27,120,27.4,0.515,21,0
5,5,109,62,41,129,35.8,0.514,25,1
7,1,140,74,26,180,24.1,0.828,23,0
...,...,...,...,...,...,...,...,...,...
609,3,96,56,34,115,24.7,0.944,39,0
610,2,96,68,13,49,21.1,0.647,26,0
611,7,102,74,40,105,37.2,0.204,45,0
612,12,100,84,33,105,30.0,0.488,46,0


In [22]:
df_train.loc[df_train["Pregnancies"] == 0]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
4,0,126,86,27,120,27.4,0.515,21,0
13,0,102,78,40,90,34.5,0.238,24,0
17,0,102,64,46,78,40.6,0.496,21,0
25,0,102,86,17,105,29.3,0.695,27,0
30,0,162,76,56,100,53.2,0.759,25,1
...,...,...,...,...,...,...,...,...,...
591,0,126,84,29,215,30.7,0.520,24,0
593,0,117,0,0,0,33.8,0.932,44,0
599,0,101,64,17,0,21.0,0.252,21,0
607,0,177,60,29,478,34.6,1.072,21,1


In [23]:
df_train.loc[(df_train["BloodPressure"] >= 70) & (df_train["BloodPressure"] <= 100)]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
1,5,166,72,19,175,25.8,0.587,51,1
2,3,96,78,39,0,37.3,0.238,40,0
3,8,110,76,0,0,27.8,0.237,58,0
4,0,126,86,27,120,27.4,0.515,21,0
6,5,97,76,27,0,35.6,0.378,52,1
...,...,...,...,...,...,...,...,...,...
602,8,188,78,0,0,47.9,0.137,43,1
603,4,173,70,14,168,29.7,0.361,33,1
608,0,131,88,0,0,31.6,0.743,32,1
611,7,102,74,40,105,37.2,0.204,45,0


In [24]:
df_train.loc[(df_train["Age"] > 50) & (df_train["BMI"] >= 25) & (df_train["BMI"] <= 30)]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
1,5,166,72,19,175,25.8,0.587,51,1
3,8,110,76,0,0,27.8,0.237,58,0
26,1,135,54,0,0,26.7,0.687,62,0
40,8,194,80,0,0,26.1,0.551,67,0
136,4,132,86,31,0,28.0,0.419,63,0
144,6,166,74,0,0,26.6,0.304,66,0
213,7,136,74,26,135,26.0,0.647,51,0
228,0,105,84,0,0,27.9,0.741,62,1
264,10,162,84,0,0,27.7,0.182,54,0
357,2,108,80,0,0,27.0,0.259,52,1


In [25]:
df_train[(df_train["Age"] >= 20) & (df_train["Age"] <= 30)].shape[0]

339

In [26]:
df_train.loc[df_train["Outcome"] == 1]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,10,108,66,0,0,32.4,0.272,42,1
1,5,166,72,19,175,25.8,0.587,51,1
5,5,109,62,41,129,35.8,0.514,25,1
6,5,97,76,27,0,35.6,0.378,52,1
10,1,168,88,29,0,35.0,0.905,52,1
...,...,...,...,...,...,...,...,...,...
602,8,188,78,0,0,47.9,0.137,43,1
603,4,173,70,14,168,29.7,0.361,33,1
607,0,177,60,29,478,34.6,1.072,21,1
608,0,131,88,0,0,31.6,0.743,32,1


In [ ]:
# part2

In [27]:
# تمرین 1
df_train.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

هیچ  سطر و ستون هایی ناشناخته ای نداریم

In [33]:
# تمرین 2
df_train.isnull().sum().sum()

np.int64(0)

In [34]:
# تمرین 3
df_train.dropna(axis=0, inplace=True)

In [36]:
# تمرین 4
df_train = df_train.reset_index(drop=True)

In [37]:
# تمرین 5
df_train.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [38]:
# تمرین 6
df_train.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [39]:
# تمرین 7
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               614 non-null    int64  
 1   Glucose                   614 non-null    int64  
 2   BloodPressure             614 non-null    int64  
 3   SkinThickness             614 non-null    int64  
 4   Insulin                   614 non-null    int64  
 5   BMI                       614 non-null    float64
 6   DiabetesPedigreeFunction  614 non-null    float64
 7   Age                       614 non-null    int64  
 8   Outcome                   614 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 43.3 KB


In [ ]:
# part3

In [40]:
# تمرین 1
(df_train == 0).sum()

Pregnancies                  94
Glucose                       5
BloodPressure                30
SkinThickness               178
Insulin                     297
BMI                           9
DiabetesPedigreeFunction      0
Age                           0
Outcome                     389
dtype: int64

In [ ]:
# تمرین 2
# Glucose
# BloodPressure
# SkinThickness
# Insulin
# BMI

In [45]:
# تمرین 3
zero_cols = [
    "Glucose",
    "BloodPressure",
    "SkinThickness",
    "Insulin",
    "BMI"
]

In [49]:
# تمرین 4
for col in zero_cols:
    median_value = df_train[col].median()

    df_train[col] = df_train[col].replace(0, median_value)

In [47]:
# تمرین 5
(df_train[zero_cols] == 0).sum()

Glucose          0
BloodPressure    0
SkinThickness    0
Insulin          0
BMI              0
dtype: int64

In [48]:
# تمرین 6
(df_train[zero_cols] == 0).sum()

Glucose          0
BloodPressure    0
SkinThickness    0
Insulin          0
BMI              0
dtype: int64